OrbitGl/App.cpp

```diff
@@ -176,15 +176,75 @@ void GLoadPdbAsync( const std::vector<std::wstring> & a_Modules )
     GModuleManager.LoadPdbAsync( a_Modules, [](){ GOrbitApp->OnPdbLoaded(); } );
 }
 
+//TODO: find a better name
+//-----------------------------------------------------------------------------
+void OrbitApp::StartRemoteCaptureBufferingThread()
+{
+    PRINT_FUNC;
+    m_TimerBuffer.clear();
+    m_SamplingCallstackBuffer.clear();
+    m_HashedSamplingCallstackBuffer.clear();
+    m_ContextSwitchBuffer.clear();
+    m_MessageBufferThread = std::make_shared<std::thread>(&OrbitApp::ProcessBufferedCaptureData, this);
+}
+
+//-----------------------------------------------------------------------------
+void OrbitApp::StopRemoteCaptureBufferingThread()
+{
+    PRINT_FUNC;
+    if ( m_MessageBufferThread )
+    {
+        m_MessageBufferThread->join();
+        m_MessageBufferThread = nullptr;
+    }
+}
+
+//-----------------------------------------------------------------------------
+void OrbitApp::ProcessBufferedCaptureData()
+{
+    while (Capture::IsCapturing())
+    {  
+        usleep(20000);
+        ScopeLock lock( m_MessageBufferMutex );
+        // timers:
+        if (m_TimerBuffer.size() > 0) {
+            Message Msg(Msg_RemoteTimers);
+            Msg.m_Size = sizeof(Timer) * m_TimerBuffer.size();
+            GTcpServer->Send(Msg, (void*)m_TimerBuffer.data());
+            m_TimerBuffer.clear();
+        }
+        
+        // sampling callstacks
+        if (m_SamplingCallstackBuffer.size() > 0){
+            std::string messageData = SerializeObjectHumanReadable(m_SamplingCallstackBuffer);
+            GTcpServer->Send(Msg_SamplingCallstacks, messageData.c_str(), messageData.size());
+            m_SamplingCallstackBuffer.clear();
+        }
+
+        // hashed sampling callstacks
+        if (m_HashedSamplingCallstackBuffer.size() > 0){  
+            std::string messageData = SerializeObjectHumanReadable(m_HashedSamplingCallstackBuffer);
+            GTcpServer->Send(Msg_SamplingHashedCallstacks, messageData.c_str(), messageData.size());
+            m_HashedSamplingCallstackBuffer.clear();
+        }
+
+        // context switches
+        if (m_ContextSwitchBuffer.size() > 0){
+            Message Msg(Msg_RemoteContextSwitches);
```

Comment:
template<class T> void TcpEntity::Send( MessageType a_Type, const std::vector<T> & a_Vector ) could make the code simpler.

GTcpServer->Send(Msg_RemoteContextSwitches, m_ContextSwitchBuffer);

instead of 

Message Msg(Msg_RemoteContextSwitches);
Msg.m_Size = sizeof(ContextSwitch) * m_ContextSwitchBuffer.size();
GTcpServer->Send(Msg, (void*)m_ContextSwitchBuffer.data());

Reply:
Thanks for the tip. But it would be obsolete if we also use the binary based serialization for context-switches.